### ISSUE TITLE
Discuss: Split omrcore into startup, RAS, and base libraries

### ISSUE THREAD

**Issue by rwy0717** [eclipse/omr#2497]
I think the responsibilities of omr core are too far reaching, and results in a web of interdependencies between components. My suggestion is to split omr core into three libraries:

1. `omrbase`: The core library would contain the basic data structures all OMR components rely on, like OMR_VM and OMR_VMThread. `-I include_core/` would be provided by omrbase. omrbase has no (or minimal) dependencies. If we want to call this library `omrcore`, people should be aware that autotools tries to delete files named `core`.

2. `omrstartup`: The startup library would provide the monolithic startup/shutdown helpers. This library depends on all other OMR components. Breaking out startup into a separate library removes core's circular dependency on other components. Further, this means that startup can be an optional library for OMR consumers. If startup also initializes the JIT, than the GC will inherit omrcore's dependency on the JIT.

3. `omrras`: Lastly, it makes sense that RAS should be it's own component, because RAS actually has it's own configure option and glue code, and I think is important enough for it's own directory.

see also:
- #2481
- #1644

/cc @dnakamura @samolisov @charliegracie

**Comment by samolisov** [comment]
Colleagues, what does RAS means?

@rwy0717 I agree, omrcore has many dependencies (OMR_TRACE_LIB, OMR_HOOK_LIB, even GC when OMR_GC is enabled, JIT in the future, etc.) for now and OMR looks like a monolithic application. I think we can go on the LLVM way: build OMR as a set of (mostly) independent libraries (as it is written in the short description of the project) to give developers an opportunity to pick only those that required.

**Comment by rwy0717** [comment]
Just checking, the core tracing engine stuff belongs in omrbase, it's the TI and RAS code that moves to omrras?

**Comment by charliegracie** [comment]
That is the hard distinction.  Technically tracing is a RAS feature but ALL components pretty much use the trace engine so if you put it in omrbase all components will require omrbase.  If you put it in RAS all components will require RAS.  I am not sure which is better.  I kind of feel like RAS is a component like util which could be required by all components.

**Comment by rwy0717** [comment]
Right. Will keep it in RAS. After looking a little more at the API component, what do you think of calling the library `omrvm`?

**Comment by rwy0717** [comment]
`omr.h` defines a large set of functions that I don't consider part of the "base" omr library. For example, `omr_agent_create` (RAS), or `omr_vm_postForkChild ` (VM, requires GC). The source files seem to be better organized. As a first pass, would we be OK to split the sources into three libraries, but leave `omr.h` for a later PR?
